[SPARK-27488][CORE] Driver interface to support GPU resources

tgravescs opened the pull request:
## What changes were proposed in this pull request?

Added the driver functionality to get the resources. 

The user interface is: SparkContext.resources  - I called it this to match the TaskContext.resources api proposed in the other PR. Originally it was going to be called SparkContext.getResources but changed to be consistent, if people have strong feelings I can change it. 

There are 2 ways the driver can discover what resources it has. 
  1) user specifies a discoveryScript, this is similar to the executors and is meant for yarn and k8s where they don't tell you what you were allocated but you are running in isolated environment.
  2) read the config spark.driver.resource.resourceName.addresses.  The config is meant to be used with standalone mode where the Worker will have to assign what GPU addresses the Driver is allowed to use by setting that config.

When the user runs a spark application, if they want the driver to have GPU's they would specify the conf spark.driver.resource.gpu.count=X  where x is the number they want.  If they are running on yarn or k8s they will also have to specify the discoveryScript as specified above, if they are on standalone mode and cluster is setup properly they wouldn't have to specify anything else.  We could potentially get rid of the spark.driver.resources.gpu.addresses config which is really meant to be an internal config for worker to set if the standalone mode Worker wanted to write a discoveryScript out and set that for the user.  I'll wait for the jira that implements that to decide if we can remove.

- This PR also has changes to be consistent about using resourceName everywhere.
- change the config names from POSTFIX to SUFFIX to be more consistent with other areas in Spark
- Moved the config checks around a bit since now used by both executor and driver. Note those might overlap a bit with https://github.com/apache/spark/pull/24374 so we will have to figure out which one should go in first.

## How was this patch tested?

Unit tests and manually test the interface.

tgravescs committed:
[SPARK-27488] Driver interface to support GPU resources

tgravescs committed:
cleanup tests

tgravescs committed:
remove extra line

tgravescs committed:
remove extra ()

SparkQA commented:
**[Test build #105423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105423/testReport)** for PR 24615 at commit [`2a8533c`](https://github.com/apache/spark/commit/2a8533c9b668d672e66e917d03e55958aaea2ccb).

tgravescs commented:
@mengxr @jiangxb1987  This might overlap a bit with https://github.com/apache/spark/pull/24374/ just let me know which way you want to resolve that.

SparkQA commented:
**[Test build #105423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105423/testReport)** for PR 24615 at commit [`2a8533c`](https://github.com/apache/spark/commit/2a8533c9b668d672e66e917d03e55958aaea2ccb).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105423/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10686/
Test PASSed.

tgravescs committed:
fix naming of getAllWithPrefixAndPostfix

tgravescs committed:
fix javadoc indentation

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10687/
Test PASSed.

SparkQA commented:
**[Test build #105424 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105424/testReport)** for PR 24615 at commit [`4ef7cb3`](https://github.com/apache/spark/commit/4ef7cb31fa8974df2fb118b6d6f4c7210b88971e).

mengxr commented:
#24374 doesn't need to handle driver resources. So I suggest removing the driver part from #24374 and work on two PRs in parallel.

SparkQA commented:
**[Test build #105424 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105424/testReport)** for PR 24615 at commit [`4ef7cb3`](https://github.com/apache/spark/commit/4ef7cb31fa8974df2fb118b6d6f4c7210b88971e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105424/
Test PASSed.

tgravescs commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -360,6 +363,47 @@ class SparkContext(config: SparkConf) extends Logging {
+    // allDriverResourceConfs is in the format Map("gpu.addresses" -> "1,2,3"),
+    // get just the resource name "gpu" and the addresses "1,2,3"
+    val resourcesWithAddrsInConfs = allDriverResourceConfs.filter {
+      case (rConf, _) => rConf.contains(SPARK_RESOURCE_ADDRESSES_POSTFIX)
```
just realized this was inconsistent with belo, one is contains, one with endsWith, I'll make that consistent

tgravescs committed:
Make utility functions in SparkConf for parsing configs and change
the name of the POSTFIX configs to be SUFFIX to be more consistent with
other things in Spark

tgravescs committed:
change resourceType to resourceName

SparkQA commented:
**[Test build #105459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105459/testReport)** for PR 24615 at commit [`b53b87a`](https://github.com/apache/spark/commit/b53b87ab38aba028f1279563646e813b3cff864e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10721/
Test PASSed.

SparkQA commented:
**[Test build #105459 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105459/testReport)** for PR 24615 at commit [`b53b87a`](https://github.com/apache/spark/commit/b53b87ab38aba028f1279563646e813b3cff864e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105459/
Test PASSed.

tgravescs commented:
@mengxr @jiangxb1987  this is ready to review when you have time.

tgravescs committed:
Add in documentation for driver configs

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10762/
Test PASSed.

SparkQA commented:
**[Test build #105504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105504/testReport)** for PR 24615 at commit [`a06967d`](https://github.com/apache/spark/commit/a06967d60fac6b5710a514b31f8fb4ac18249f77).

SparkQA commented:
**[Test build #105504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105504/testReport)** for PR 24615 at commit [`a06967d`](https://github.com/apache/spark/commit/a06967d60fac6b5710a514b31f8fb4ac18249f77).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105504/
Test PASSed.

tgravescs committed:
Add in logging messages for Driver resources

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10881/
Test PASSed.

SparkQA commented:
**[Test build #105628 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105628/testReport)** for PR 24615 at commit [`e3694f3`](https://github.com/apache/spark/commit/e3694f317b4614010d86fa14fb15297784d8f196).

SparkQA commented:
**[Test build #105628 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105628/testReport)** for PR 24615 at commit [`e3694f3`](https://github.com/apache/spark/commit/e3694f317b4614010d86fa14fb15297784d8f196).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105628/
Test PASSed.

mengxr commented:
making a pass:)

mengxr commented on core/src/main/scala/org/apache/spark/ResourceDiscoverer.scala:
```diff
@@ -42,28 +42,40 @@ private[spark] object ResourceDiscoverer extends Logging {
+  def discoverResourcesInformation(
+      sparkConf: SparkConf,
+      confPrefix: String,
+      resources: Option[Set[String]] = None
```
`resourceNamesOpt: optionally specify resource names. if not set ...`

tgravescs replied:
will update

tgravescs replied:
pushed

mengxr commented on core/src/main/scala/org/apache/spark/ResourceDiscoverer.scala:
```diff
@@ -73,21 +85,45 @@ private[spark] object ResourceDiscoverer extends Logging {
     result
   }
+
+  // Make sure the actual resources we have on startup is at least the number the user
```
Can we make it ScalaDoc?

tgravescs replied:
yep

mengxr commented on core/src/main/scala/org/apache/spark/ResourceDiscoverer.scala:
```diff
@@ -73,21 +85,45 @@ private[spark] object ResourceDiscoverer extends Logging {
+  // for each task resource requirement and that they are large enough. This function
+  // is used by both driver and executors.
+  def checkActualResourcesMeetRequirements(
+      requiredResources: Map[String, String],
```
not in this PR, it would be nice if we use a class to capsulate the resource request

mengxr commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -406,6 +406,13 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
+  /**
+   * Get all parameters that start with `prefix` and end with 'suffix'
+   */
+  def getAllWithPrefixAndSuffix(prefix: String, suffix: String): Array[(String, String)] = {
```
I saw this method is frequently used but I'm not sure this is the right home for it. Even `object SparkConf` is package private, to me those methods seem too specific and they will show up in Java API docs. The return result is also very specific to our use case. cc: @gatorsmile for suggestions

tgravescs replied:
I put it here because it really seems like just a variation of getAllWithPrefix. There are a bunch of configs now that use this prefix type format, but if you have a better idea I can move it and we can always propogate it up if some one else needs it.

mengxr commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -587,6 +594,30 @@ class SparkConf(loadDefaults: Boolean) extends Cloneable with Logging with Seria
+            "needs to be specified since a task requirement config: " +
+            s"${SPARK_TASK_RESOURCE_PREFIX + rName + SPARK_RESOURCE_COUNT_SUFFIX} was specified")
+      )
+      if (execCount.toLong < taskCount.toLong) {
```
Shall we produce a warning if the executor requested count is not a multiple of task count?

tgravescs replied:
that should be in the scheduler PR

mengxr commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -777,6 +808,35 @@ private[spark] object SparkConf extends Logging {
+   * passed in and returns just the base part of the config before the first
+   * '.' and its value.
+   */
+  def getConfigsWithSuffix(
```
Ditto. This is also too specific. Looking at the method name,  I cannot guess what the method does, in particular, it also change the key in the return result.

tgravescs replied:
ok what do you suggest?

tgravescs replied:
I can just move them into the Resourcediscoverer object and try to rename or make a ResourceConfUtils class, would like your input before changing so we don't go back and forth a bunch of times

mengxr commented on core/src/main/scala/org/apache/spark/SparkConf.scala:
```diff
@@ -777,6 +808,35 @@ private[spark] object SparkConf extends Logging {
+   * getAllWithPrefix function and returns the base part of the config
+   * before the first '.'.
+   */
+  def getBaseOfConfigs(configs: Array[(String, String)]): Set[String] = {
```
Ditto.

tgravescs replied:
again what do you suggest then?

mengxr commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -360,6 +363,48 @@ class SparkContext(config: SparkConf) extends Logging {
+    val resourcesWithAddrsInConfs =
+      SparkConf.getConfigsWithSuffix(allDriverResourceConfs, SPARK_RESOURCE_ADDRESSES_SUFFIX)
+
+    _resources = if (resourcesWithAddrsInConfs.nonEmpty) {
```
It is possible that some resources have addresses passed in while some other resources require discovery script. So it is not one or the other.

tgravescs replied:
in what case are you thinking?   The address config is really an internal config meant for standalone mode.  
The only case I can think of is if a user requests a resource the standalone mode worker handles but then asks for another resource that it doesn't and thus needs discoveryScript.   But that means that they are on their own for handling isolation and assignment.

tgravescs replied:
also note this matches the way the executor does it as well. For the executor, either the worker in standalone mode passes in a resources file with what it has allocated per executor or it doesn't pass a file and we use the discovery scripts.  Yarn and k8s I don't expect to have issues with this as they will always use discovery Script.  If we want to be able to support both ways we should update both places.   I don't run standalone mode much so if you have a usecase this makes sense we can certainly change it.

mengxr replied:
It is fine to have that assumption. But we should check and throw when `resourcesWithAddrsInConfs.nonEmpty` but there are remaining resources that require discovery. In the current implementation, it is silently ignored.

tgravescs replied:
we can support both ways, lets do this with the abstraction layer changes though

mengxr commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -83,51 +83,11 @@ private[spark] class CoarseGrainedExecutorBackend(
-    val taskResourceConfigs = env.conf.getAllWithPrefix(SPARK_TASK_RESOURCE_PREFIX)
-    val resourceInfo = if (taskResourceConfigs.nonEmpty) {
-      val execResources = resourcesFile.map { resourceFileStr => {
+    val resourceInfo = if (env.conf.getAllWithPrefix(SPARK_TASK_RESOURCE_PREFIX).nonEmpty) {
```
`SPARK_EXECUTOR_RESOURCE_PREFIX`?

tgravescs replied:
No, I only want to look for resources is a task we are running requires them.  If no tasks require other resources, then its just overhead for us.

tgravescs committed:
Update comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10889/
Test PASSed.

SparkQA commented:
**[Test build #105635 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105635/testReport)** for PR 24615 at commit [`c6d513c`](https://github.com/apache/spark/commit/c6d513cb050074a66e2497bf4c7242daa25755c1).

SparkQA commented:
**[Test build #105635 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105635/testReport)** for PR 24615 at commit [`c6d513c`](https://github.com/apache/spark/commit/c6d513cb050074a66e2497bf4c7242daa25755c1).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105635/
Test FAILed.

tgravescs committed:
update scaladoc

tgravescs committed:
Update scala doc again

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10891/
Test PASSed.

SparkQA commented:
**[Test build #105637 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105637/testReport)** for PR 24615 at commit [`f41e146`](https://github.com/apache/spark/commit/f41e14690c0714bc8c5f35b41c333f3e830cd9e3).

mengxr commented:
@tgravescs About the conf utilities, I think it hurts readability because we didn't have the abstraction at resource level. We went to manipulate confs directly. I have a proposal (not tested against your code) below. Let me know how you like the proposal. We can also merge the PR in current form and do the refactoring in a later PR.

~~~scala
/**
 *
 * @param componentName spark.driver / spark.executor / spark.task
 * @param resourceName gpu, fpga
 */
private[spark] case class ResourceID(componentName: String, resourceName: String) {
  def confPrefix: String = s"$componentName.resource.$resourceName." // with ending dot
}

private[spark] case class ResourceRequest(id: ResourceID, count: Double, discoveryScript: String)

// We can define toResourceInfo() to map it to a public ResourceInformation instance.
private[spark] case class ResourceAllocation(id: ResourceID, addresses: Seq[String])

object ResourceUtils {

  val COUNT = "count"
  val DISCOVERY_SCRIPT = "discoveryScript"
  val ADDRESSES = "addresses"

  def parseRequest(sparkConf: SparkConf, resourceId: ResourceID): ResourceRequest = {
    val settings = sparkConf.getAllWithPrefix(resourceId.confPrefix).toMap
    val quantity = settings(COUNT).toDouble
    val discoveryScript = settings(DISCOVERY_SCRIPT)
    ResourceRequest(resourceId, quantity, discoveryScript)
  }

  def listResourceIds(sparkConf: SparkConf, componentName: String): Seq[ResourceID] = {
    sparkConf.getAllWithPrefix(s"$componentName.resource.").map { case (key, _) =>
      key.substring(0, key.indexOf('.'))
    }.toSet.toSeq.map(name => ResourceID(componentName, name))
  }

  def parseAllRequests(
      sparkConf: SparkConf, componentName: String): Seq[ResourceRequest] = {
    listResourceIds(sparkConf, componentName).map { id =>
      parseRequest(sparkConf, id)
    }
  }

  def parseResourceAllocationJson(resources: String): Seq[ResourceAllocation] = {
    implicit val formats = DefaultFormats
    parse(resources).extract[Seq[ResourceAllocation]]
  }

  def discoverResource(resourceRequest: ResourceRequest): ResourceAllocation = ???

  def parseAllocatedOrDiscoverResources(
      sparkConf: SparkConf,
      componentName: String,
      resourcesOpt: Option[String]): Seq[ResourceAllocation] = {
    val allocated = resourcesOpt.flatMap(parseResourceAllocationJson(_))
      .filter(_.id.componentName == componentName)
      .toSeq
    val otherResourceIds = listResourceIds(sparkConf, componentName).diff(allocated.map(_.id))
    allocated ++ otherResourceIds.map { id =>
      val request = parseRequest(sparkConf, id)
      discoverResource(request)
    }
  }

  def assertResourceAllocationMeetsRequest(
      allocation: ResourceAllocation, request: ResourceRequest): Unit = {
    require(allocation.id == request.id && allocation.addresses.size >= request.count)
  }

  def assertAllResourceAllocationMeetRequests(
      allocations: Seq[ResourceAllocation], requests: Seq[ResourceRequest]): Unit = {
    val allocated = allocations.map(x => x.id -> x).toMap
    requests.foreach { r =>
      assertResourceAllocationMeetsRequest(allocated(r.id), r)
    }
  }

  // also add some methods to convert requests to Spark conf entries to simplify tests
}
~~~

SparkQA commented:
**[Test build #105637 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105637/testReport)** for PR 24615 at commit [`f41e146`](https://github.com/apache/spark/commit/f41e14690c0714bc8c5f35b41c333f3e830cd9e3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105637/
Test PASSed.

tgravescs commented:
Sure, we can add in the extra abstraction for some of the resources, let me do it here as it changes things a bit and would rather do it sooner then later to avoid changing things yet again

jiangxb1987 reviewed (commented):
This PR looks good from my side, the only concern is if we are to add the abstraction for Resource layer here? If we do then it means #24374 need to rebase on this PR, if we do the abstraction in a separated PR then the rebase work may be easier. Again this PR looks in a good sharp from my side, so I'm thinking whether we shall merge this PR first and then refactor with an extra abstraction Resource layer, but I'm also not against the other way.

mengxr commented:
I'm okay with merging the PR first and doing the refactor later, which I can take.

tgravescs commented on docs/configuration.md:
```diff
@@ -187,6 +187,25 @@ of the most common options to set are:
+  <td>
+    The number of a particular resource type to use on the driver.
+    If this is used, you must also specify the
+    <code>spark.driver.resource.{resourceName}.discoveryScript</code>
```
addresses is meant to be an internal config that standalone mode sets so I don't want to publish for users.

tgravescs commented:
Its up to you, I have the code rewritten to add the abstraction layer, I'm just finishing the tests, but like you said the scheduler PR would have to be rebased.

I was also changing the driver config from .addresses to specify a .resourceFile so that it matches the executor side and makes the parsing code common.

Again both of which can be done separately but we should decide on the order we want things to be merged.

I'll push the changes for the 2 comments above that need changes here first and wait to hear on what we want to do.

tgravescs commented:
Actually I had already pushed the 2 changes to the comments and naming that were suggested, everything else I believe falls into the abstraction layer.

tgravescs committed:
Merge branch 'master' of https://github.com/apache/spark into SPARK-27488-orig

SparkQA commented:
**[Test build #105728 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105728/testReport)** for PR 24615 at commit [`768392c`](https://github.com/apache/spark/commit/768392ca83c29a2ef6b3ecb2d175cccd20ffa0ab).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/10984/
Test PASSed.

SparkQA commented:
**[Test build #105728 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105728/testReport)** for PR 24615 at commit [`768392c`](https://github.com/apache/spark/commit/768392ca83c29a2ef6b3ecb2d175cccd20ffa0ab).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105728/
Test PASSed.

tgravescs commented:
@mengxr  for the standalone mode, the executor has the resourceFile it passes in and I'm looking at making the driver have the same thing (which again we can do separate depending on what we decide), do you want it to be able to have the "component name" in the json (that is what your code above implies)?  I don't see a reason for it unless you are wanting to have a single file that has both the driver and the executor allocations.

mengxr commented:
I prefer keeping the component name, but no strong preference. Having the component name would help uniquely identify the resource, and hence simplify the code that looks up back at the Spark conf.

mengxr commented:
Let's merge this PR and Xingbo's then do the refactoring.

mengxr commented:
Merged into master. Thanks!
